### PR TITLE
Further changes in relation to #19171

### PR DESCRIPTION
**Pull request by jonleighton**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19272)

### CODE REVIEW

**Comment by jonleighton** [comment]
I don't know if this `TextControl` trait is the right abstraction, so feedback on that front is very welcome. It kind of annoys me that now there's lots of boilerplate code to call in to the methods defined by `TextControl`.

**Comment by KiChjang** [comment]
This is a larger change than I expected, and I'm not comfortable reviewing this. r? @jdm

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-start-end.html [hunk 1]
```diff
@@ -59,10 +59,18 @@
     assert_equals(testValue.length, 10);
   }, "Sanity check for testValue length; if this fails, variou absolute offsets in the test below need to be adjusted to be less than testValue.length");
 
+  for (let prop of ["selectionStart", "selectionEnd"]) {
+    test(function() {
```
Comment (jdm) [inline tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-start-end.html]: This outer test is not necessary.

**Review by jdm — CHANGES_REQUESTED** [review]
This seems like a sensible solution. The only other abstraction I came up with is free functions that accept the textinput as an argument, rather than putting them all in a trait, but I don't see a compelling reason to use that design instead.

**Comment by jdm** [comment]
Also `./mach test-tidy` fails with these changes.

File: components/script/dom/textcontrol.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+use dom::node::{Node, NodeDamage, window_from_node};
+use script_traits::ScriptToConstellationChan;
+use textinput::{SelectionDirection, TextInput};
+
+pub trait TextControl: DerivedFrom<EventTarget> + DerivedFrom<Node> {
+    fn textinput(&self) -> &DomRefCell<TextInput<ScriptToConstellationChan>>;
+
+    // https://html.spec.whatwg.org/multipage/#dom-textarea/input-selectionstart
+    fn dom_selection_start(&self) -> u32 {
+        self.textinput().borrow().get_selection_start()
```
Comment (KiChjang) [inline components/script/dom/textcontrol.rs]: When I mentioned that whether the selection{Start, End, Range} functions matches the spec, I meant whether they are following the steps as defined in the spec. This function in particular does not seem to have included step 1 and 2.

**Comment by KiChjang** [comment]
Also, may I ask why this abstraction is necessary? It seems to me that the intent is to generalize the textinput stuff across HTMLInputElement and HTMLTextAreaElement, but I may be wrong on that.

**Comment by jonleighton** [comment]
> The only other abstraction I came up with is free functions that accept the textinput as an argument

Out of curiosity, can you point me to anything which describes what is meant by a "free function"? (I'm not very familiar with Rust yet...)

> Also ./mach test-tidy fails with these changes.

I'll look into that.

> When I mentioned that whether the selection{Start, End, Range} functions matches the spec, I meant whether they are following the steps as defined in the spec. This function in particular does not seem to have included step 1 and 2.

Yeah, I'm not done yet but I thought it was worth submitting this PR to get feedback before I make further changes. For now, the implementation of `TextControl` is simply moved from elsewhere in the code base. I'd then intend to work through the spec to make sure it conforms to everything. (I can continue working on the subsequent changes if desired, but I just wanted to get some feedback so I don't go too far down an undesirable path.)

> Also, may I ask why this abstraction is necessary? It seems to me that the intent is to generalize the textinput stuff across HTMLInputElement and HTMLTextAreaElement, but I may be wrong on that.

Yes, that's the intent. The spec of the selection API is identical for both `input` and `textarea`, but until this change they both had separate implementations in the Servo codebase. However, the separate implementations were very similar and had clearly been copy/pasted. When I fixed the firing of the `select` event on assignment to `selection{Start,End}` for `input`, my fix didn't apply to `textarea`. To fix it for `textarea` would require me to make an identical change to `HTMLTextAreaElement`. It seemed more sensible to extract the common code so that the two elements behave the same way and there is no duplication. (This extraction implicitly fixes the implementation for `textarea` hence some more tests now pass.)

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-start-end.html (hunk 1 above)
Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-start-end.html]: Thanks. I thought that maybe it was a bad idea to call `createTestElements` outside of a `test`, but perhaps it doesn't actually matter.

**Comment by KiChjang** [comment]
A free function is a function that isn't inside an impl block, i.e. it exists at the top level of a module.

**Comment by jonleighton** [comment]
Also ran `./mach update-manifest` (what is the `MANIFEST.json` file for?)

**Comment by KiChjang** [comment]
So I thought a bit about this, and was wondering whether the trait abstraction is necessary. It's true that HTMLInputElement and HTMLTextAreaElement don't share the selection code, but they both do have a `textinput` field in their structs. I'm wondering if the trait methods you define should better be located there.

**Comment by jonleighton** [comment]
Yeah, I thought about that too. The reason I didn't go down that path is that `TextInput` wouldn't be able to handle anything DOM-related, i.e. firing events, marking the node dirty, etc. As the selection spec is further implemented, I expect there to be more common DOM-related code than there currently is, so it seems like a good idea to have an abstraction which can cover that stuff.

**Comment by KiChjang** [comment]
I think it's actually doable to include DOM-related stuff as parameters to functions defined in `textinput.rs`, e.g. for `set_selection_range`, you can pass in a `Window` and `Node` parameter in order for it to performs its job.

**Comment by jdm** [comment]
@bors-servo: r+
I am fine with the current abstraction. We can revisit the design as further changes are made.

**Pull request merged**